File: task2.py
def find_unique(v):
    unique = []
    for email in v:
        if email not in unique:
            unique.append(email)
    return unique

def format_email(s):
    result = []
    seen_at = False
    seen_star = False
    
    for char in s:
        if char == '@':
            seen_at = True
        if not seen_at:
            if char == '*':
                seen_star = True
                continue
            if seen_star:
                continue
            if char == '.':
                continue
        result.append(char)
    
    return ''.join(result)

def is_allowed(s):
    not_allowed_chars = {'&', '=', '+', '-', '_', '<', '>', ',', "'"}
    
    if not s or s[0] == '.':
        return False
    if '@' not in s:
        return False
    
    at_index = s.index('@')
    if at_index < 6 or at_index > 30:
        return False
    
    for char in s[:at_index]:
        if char in not_allowed_chars:
            return False
    
    if s.rfind('.') <= at_index or '.' not in s:
        return False
    
    return True

def main():
    emails = [".555555@.", "123456*7890@.mailru", "123456@.mailru", 
              "123456@.mailru", "1234567@.mailru"]
    
    # Фильтрация и форматирование email
    valid_emails = [format_email(email) for email in emails if is_allowed(email)]
    
    if not valid_emails:
        print("No unique emails")
    else:
        print(len(find_unique(valid_emails)))

if __name__ == "__main__":
    main()

File: task2.rb
def find_unique(v)
  v.uniq
end

def format_email(s)
  result = []
  seen_at = false
  seen_star = false
  
  s.each_char do |c|
    if c == '@'
      seen_at = true
    end
    unless seen_at
      if c == '*'
        seen_star = true
        next
      end
      if seen_star
        next
      end
      if c == '.'
        next
      end
    end
    result << c
  end
  result.join
end

def is_allowed(s)
  not_allowed_chars = ['&', '=', '+', '-', '_', '<', '>', ',', "'"]
  return false if s.nil? || s.empty? || s[0] == '.'
  return false unless s.include?('@')
  
  at_index = s.index('@')
  return false if at_index < 6 || at_index > 30
  
  s[0...at_index].each_char do |c|
    return false if not_allowed_chars.include?(c)
  end
  
  last_dot = s.rindex('.')
  return false if last_dot.nil? || last_dot <= at_index
  true
end

v = [".555555@.", "1111.111@mail.ru", "11111111@.mailru", "1111111@.mailru", "11111111@.mailru"]
filtered = v.select { |email| is_allowed(email) }.map { |email| format_email(email) }
if filtered.empty?
  puts "No unique emails"
else
  puts find_unique(filtered).size
end
